app/jobs/importing/hud_zip/fetch_and_import_job.rb

```diff
@@ -7,22 +7,46 @@
 module Importing::HudZip
   class FetchAndImportJob < BaseJob
     queue_as ENV.fetch('DJ_LONG_QUEUE_NAME', :long_running)
+    WAIT_MINUTES = 15
+
+    # When you call jobs with .perform_later, they are executed in the ActiveJob world, which doesn't
+    # obey they max_attempts for Delayed Job.  We'll adjust the attempts to give us what we want
+    after_enqueue :enforce_max_attempts
 
     def perform(klass:, options:)
       safe_klass = known_classes.detect { |m| klass == m }
-      raise "Unknown import class: #{klass}; You must add it to the whitelist in FetchAndImportJob" unless safe_klass.present?
+      raise "Unknown import class: #{klass}; You must add it to the list of known classes in FetchAndImportJob" unless safe_klass.present?
 
-      safe_klass.constantize.new(**options).import!
-    end
+      data_source_id = options[:data_source_id]
+      lock_obtained = GrdaWarehouse::DataSource.with_advisory_lock(advisory_lock_name(data_source_id), timeout_seconds: 60) do
+        safe_klass.constantize.new(**options).import!
+        # To prevent re-running when called against the same files if run more than once in a day, yield true
+        true
+      end
 
-    def max_attempts
-      1
+      requeue_at(Time.current + WAIT_MINUTES.minutes, "Import of Data Source: #{data_source_id} already running...re-queuing job for #{WAIT_MINUTES} minutes from now") unless lock_obtained
     end
 
     def known_classes
       [
         'Importers::HmisAutoMigrate::S3',
       ].freeze
     end
+
+    private def advisory_lock_name(data_source_id)
+      GrdaWarehouse::DataSource.import_advisory_lock_name(data_source_id)
+    end
+
+    def enforce_max_attempts
+      delayed_job.update!(attempts: calculated_attempts)
+    end
```

Comment:
If the job calls requeue_at when it fails, is this hook still needed? It seems like attempts would already be set to calculated_attempts at that point. Or is DJ updating attempts and this ensures it will run again? Maybe add a comment as to why it's needed

Reply:
This doesn't call `requeue_at` when it fails, only when it fails to get the lock.  The `requeue_at` process allows this instance of the delayed job to complete successfully, but creates a new copy of it scheduled for the new time.

`enforce_max_attempts` is only really needed on the initial creation of the job to prevent it from running more than once.